In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('Files used in current analysis/Linkedin_Job_Details_26-Sep-2023-to-09-Oct-2023_4214-jobs_181-pages.csv')
df

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
4209         Oracle Apps DBA - R12 (Full Customization)   
4210                Engineer, Service-to-Service Fabric   
4211  Technical Product Manager (OTE $80,000/year US...   
4212                               Senior IOS Developer   
4213  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
4209  Rozee.pk · Lahore, Punjab, Pakistan  \n4 days ago   
4210  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
4211  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
4212  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
4213  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
4209                        On-site\n\n \nFull-time   
4210  Remote\n\n \nFull-time\n\n \nMid-Senior level   
4211       Remote\n\n \nFull-time\n\n \nEntry level   
4212  Remote\n\n \nFull-time\n\n \nMid-Senior level   
4213       Remote\n\n \nFull-time\n\n \nEntry level   

                                           Job Link  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
4209  https://www.linkedin.com/jobs/view/3734517227   
4210  https://www.linkedin.com/jobs/view/3730554639   
4211  https://www.linkedin.com/jobs/view/3731001750   
4212  https://www.linkedin.com/jobs/view/3736179533   
4213  https://www.linkedin.com/jobs/view/3731006178   

                                              employees  
0                                       11-50 employees  
1                                       11-50 employees  
2                                      51-200 employees  
3                                       11-50 employees  
4                   Mobeen Naqvi is hiring for this job  
...                                                 ...  
4209                                501-1,000 employees  
4210      5,001-10,000 employees · Software Development  
4211   51-200 employees · IT Services and IT Consulting  
4212  51-200 employees · Transportation, Logistics, ...  
4213   51-200 employees · IT Services and IT Consulting  

[4214 rows x 5 columns]

In [3]:
df_=pd.read_csv('Files used in current analysis/Linkedin_Job-Descriptions-company_details_26-Sep-2023-to-26-Nov-2023_15947-jobs_685-pages.csv')
df_


Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15942  https://www.linkedin.com/jobs/view/3767355025   
15943  https://www.linkedin.com/jobs/view/3766809535   
15944  https://www.linkedin.com/jobs/view/3772530977   
15945  https://www.linkedin.com/jobs/view/3771943212   
15946  https://www.linkedin.com/jobs/view/3767944844   

                                          Job Desception  \
0      About the job\n            \n\n \nAbout Us:We ...   
1      About the job\n            \n\n \nWe are looki...   
2      About the job\n            \n\n \nCompany Over...   
3      About the job\n            \n\n \nBI Developer...   
4      About the job\n            \n\n \nIt is a prog...   
...                                                  ...   
15942  About the job\n            \n \nThis role is p...   
15943  About the job\n            \n \nA U.S.-based c...   
15944                                      About the job   
15945                                      About the job   
15946                                      About the job   

                                            Company Info  
0      Advertising Services\n            \n          ...  
1      IT Services and IT Consulting\n            \n ...  
2      IT Services and IT Consulting\n            \n ...  
3      IT Services and IT Consulting\n            \n ...  
4                                                    NaN  
...                                                  ...  
15942                                                NaN  
15943  IT Services and IT Consulting\n            \n ...  
15944                                                NaN  
15945                                                NaN  
15946                                                NaN  

[15947 rows x 3 columns]

In [4]:
df_1=pd.read_csv('Files used in current analysis/Linkedin_Job_Details_26-Sep-2023-to-26-Nov-2023_15875-jobs_685-pages.csv')
df_1

Job Title  \
0                              LinkedIn Data Extractor   
1                                 It Management Intern   
2                               SQL Database Developer   
3                      Business Intelligence Developer   
4                           Web Developer - Internship   
...                                                ...   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15871                         Release Support Engineer   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
0            EcomFleet · Pakistan (Remote)  \n1 week ago   
1      Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2      Murkez Technologies · Lahore, Punjab, Pakistan...   
3      DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4      ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                  ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15871           Turing · Pakistan Reposted  \n1 week ago   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
0                                           Full-time   
1                                           Full-time   
2                                           Full-time   
3                                           Full-time   
4                                           Full-time   
...                                               ...   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15871        Remote\n\n \nFull-time\n\n \nEntry level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15870  https://www.linkedin.com/jobs/view/3767355025   
15871  https://www.linkedin.com/jobs/view/3766809535   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees  
0                                        11-50 employees  
1                                        11-50 employees  
2                                       51-200 employees  
3                                        11-50 employees  
4                    Mobeen Naqvi is hiring for this job  
...                                                  ...  
15870                   5,001-10,000 employees · Banking  
15871  1,001-5,000 employees · IT Services and IT Con...  
15872   51-200 employees · IT Services and IT Consulting  
15873  201-500 employees · Outsourcing and Offshoring...  
15874                                    11-50 employees  

[15875 rows x 5 columns]

In [5]:
df['Job ID']=df['Job Link'].apply(lambda x:int(x.split('/')[5]))
df_['Job ID']=df_['Job Link'].apply(lambda x:int(x.split('/')[5]))
df_1['Job ID']=df_1['Job Link'].apply(lambda x:int(x.split('/')[5]))

In [6]:
df_2=df_[df_['Job ID'].isin(df['Job ID'])]
df_2

Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15679  https://www.linkedin.com/jobs/view/3729723124   
15681  https://www.linkedin.com/jobs/view/3537348803   
15766  https://www.linkedin.com/jobs/view/3661745501   
15776  https://www.linkedin.com/jobs/view/3728757956   
15915  https://www.linkedin.com/jobs/view/3734537706   

                                          Job Desception  \
0      About the job\n            \n\n \nAbout Us:We ...   
1      About the job\n            \n\n \nWe are looki...   
2      About the job\n            \n\n \nCompany Over...   
3      About the job\n            \n\n \nBI Developer...   
4      About the job\n            \n\n \nIt is a prog...   
...                                                  ...   
15679  About the job\n            \n \nJob Responsibi...   
15681  About the job\n            \n \nLOCATION: Jara...   
15766  About the job\n            \n \nAbout Us:CureM...   
15776  About the job\n            \n \nAbout the role...   
15915  About the job\n            \n \nPosition Title...   

                                            Company Info      Job ID  
0      Advertising Services\n            \n          ...  3727054368  
1      IT Services and IT Consulting\n            \n ...  3723395858  
2      IT Services and IT Consulting\n            \n ...  3727034665  
3      IT Services and IT Consulting\n            \n ...  3727029273  
4                                                    NaN  3727034696  
...                                                  ...         ...  
15679  Software Development\n            \n          ...  3729723124  
15681                                                NaN  3537348803  
15766  Software Development\n            \n          ...  3661745501  
15776  Book and Periodical Publishing\n            \n...  3728757956  
15915  Internet Marketplace Platforms\n            \n...  3734537706  

[4904 rows x 4 columns]

In [7]:
df_3=pd.merge(df,df_,on='Job ID',how='inner')
df_3

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
7267                Engineer, Service-to-Service Fabric   
7268  Technical Product Manager (OTE $80,000/year US...   
7269                               Senior IOS Developer   
7270                               Senior IOS Developer   
7271  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
7267  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
7268  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
7269  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7270  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7271  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
7267  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7268       Remote\n\n \nFull-time\n\n \nEntry level   
7269  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7270  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7271       Remote\n\n \nFull-time\n\n \nEntry level   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https://www.linkedin.com/jobs/view/3731001750   
7269  https://www.linkedin.com/jobs/view/3736179533   
7270  https://www.linkedin.com/jobs/view/3736179533   
7271  https://www.linkedin.com/jobs/view/3731006178   

                                              employees      Job ID  \
0                                       11-50 employees  3727054368   
1                                       11-50 employees  3723395858   
2                                      51-200 employees  3727034665   
3                                       11-50 employees  3727029273   
4                   Mobeen Naqvi is hiring for this job  3727034696   
...                                                 ...         ...   
7267      5,001-10,000 employees · Software Development  3730554639   
7268   51-200 employees · IT Services and IT Consulting  3731001750   
7269  51-200 employees · Transportation, Logistics, ...  3736179533   
7270  51-200 employees · Transportation, Logistics, ...  3736179533   
7271   51-200 employees · IT Services and IT Consulting  3731006178   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https:/

In [8]:
# why df3 have 7272 rows

In [9]:
# df3=df[df['Job ID']==df_['Job ID']]
# df3
# ValueError: Can only compare identically-labeled Series objects

In [10]:
df_2.drop_duplicates(subset=['Job ID']).shape


(3623, 4)

In [11]:
df.drop_duplicates(subset=['Job ID']).shape

(3623, 6)

In [12]:
df_3.drop_duplicates(subset=['Job ID']).shape

(3623, 9)

In [13]:
df.isnull().sum()

Job Title    54
Job Info     54
Job Type     54
Job Link      0
employees    54
Job ID        0
dtype: int64

In [14]:
df_3.isnull().sum()

Job Title          67
Job Info           67
Job Type           67
Job Link_x          0
employees          67
Job ID              0
Job Link_y          0
Job Desception    104
Company Info      915
dtype: int64

In [15]:
df_3.columns

Index(['Job Title', 'Job Info', 'Job Type', 'Job Link_x', 'employees',
       'Job ID', 'Job Link_y', 'Job Desception', 'Company Info'],
      dtype='object')

In [16]:
df.dropna(inplace=True)
df_1.dropna(inplace=True)
df_3.dropna(subset=['Job Title','Job Info','Job Type','employees'],inplace=True) # i want to delete specific columns

In [17]:
# df['Job ID']=df['Job Link'].apply(lambda x:int(x.split('/')[5]))

In [18]:
df = df.drop_duplicates(subset=['Job ID'])
df_1=df_1.drop_duplicates(subset=['Job ID'])
df_2=df_2.drop_duplicates(subset=['Job ID'])
df_3=df_3.drop_duplicates(subset=['Job ID'])

In [19]:
df.shape,df_1.shape,df_2.shape,df_3.shape

((3573, 6), (11744, 6), (3623, 4), (3573, 9))

In [20]:
df_2

Job Link  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
4210  https://www.linkedin.com/jobs/view/3734517227   
4211  https://www.linkedin.com/jobs/view/3730554639   
4212  https://www.linkedin.com/jobs/view/3731001750   
4213  https://www.linkedin.com/jobs/view/3736179533   
4214  https://www.linkedin.com/jobs/view/3731006178   

                                         Job Desception  \
0     About the job\n            \n\n \nAbout Us:We ...   
1     About the job\n            \n\n \nWe are looki...   
2     About the job\n            \n\n \nCompany Over...   
3     About the job\n            \n\n \nBI Developer...   
4     About the job\n            \n\n \nIt is a prog...   
...                                                 ...   
4210  About the job\n            \n\n \nJob Summary:...   
4211  About the job\n            \n\n \nOverviewWork...   
4212  About the job\n            \n\n \nAre you a hi...   
4213  About the job\n            \n\n \nAbout Ekar C...   
4214  About the job\n            \n\n \nAre you a hi...   

                                           Company Info      Job ID  
0     Advertising Services\n            \n          ...  3727054368  
1     IT Services and IT Consulting\n            \n ...  3723395858  
2     IT Services and IT Consulting\n            \n ...  3727034665  
3     IT Services and IT Consulting\n            \n ...  3727029273  
4                                                   NaN  3727034696  
...                                                 ...         ...  
4210  Business Consulting and Services\n            ...  3734517227  
4211  Software Development\n            \n          ...  3730554639  
4212  IT Services and IT Consulting\n            \n ...  3731001750  
4213  Transportation, Logistics, Supply Chain and St...  3736179533  
4214  IT Services and IT Consulting\n            \n ...  3731006178  

[3623 rows x 4 columns]

In [21]:
df_3

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
7266         Oracle Apps DBA - R12 (Full Customization)   
7267                Engineer, Service-to-Service Fabric   
7268  Technical Product Manager (OTE $80,000/year US...   
7269                               Senior IOS Developer   
7271  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
7266  Rozee.pk · Lahore, Punjab, Pakistan  \n4 days ago   
7267  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
7268  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
7269  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7271  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
7266                        On-site\n\n \nFull-time   
7267  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7268       Remote\n\n \nFull-time\n\n \nEntry level   
7269  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7271       Remote\n\n \nFull-time\n\n \nEntry level   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https://www.linkedin.com/jobs/view/3731001750   
7269  https://www.linkedin.com/jobs/view/3736179533   
7271  https://www.linkedin.com/jobs/view/3731006178   

                                              employees      Job ID  \
0                                       11-50 employees  3727054368   
1                                       11-50 employees  3723395858   
2                                      51-200 employees  3727034665   
3                                       11-50 employees  3727029273   
4                   Mobeen Naqvi is hiring for this job  3727034696   
...                                                 ...         ...   
7266                                501-1,000 employees  3734517227   
7267      5,001-10,000 employees · Software Development  3730554639   
7268   51-200 employees · IT Services and IT Consulting  3731001750   
7269  51-200 employees · Transportation, Logistics, ...  3736179533   
7271   51-200 employees · IT Services and IT Consulting  3731006178   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https:/

In [22]:
df_4=df_4=df_3.dropna(subset=['Job Desception'])
df_4

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
7266         Oracle Apps DBA - R12 (Full Customization)   
7267                Engineer, Service-to-Service Fabric   
7268  Technical Product Manager (OTE $80,000/year US...   
7269                               Senior IOS Developer   
7271  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
7266  Rozee.pk · Lahore, Punjab, Pakistan  \n4 days ago   
7267  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
7268  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
7269  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7271  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
7266                        On-site\n\n \nFull-time   
7267  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7268       Remote\n\n \nFull-time\n\n \nEntry level   
7269  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7271       Remote\n\n \nFull-time\n\n \nEntry level   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https://www.linkedin.com/jobs/view/3731001750   
7269  https://www.linkedin.com/jobs/view/3736179533   
7271  https://www.linkedin.com/jobs/view/3731006178   

                                              employees      Job ID  \
0                                       11-50 employees  3727054368   
1                                       11-50 employees  3723395858   
2                                      51-200 employees  3727034665   
3                                       11-50 employees  3727029273   
4                   Mobeen Naqvi is hiring for this job  3727034696   
...                                                 ...         ...   
7266                                501-1,000 employees  3734517227   
7267      5,001-10,000 employees · Software Development  3730554639   
7268   51-200 employees · IT Services and IT Consulting  3731001750   
7269  51-200 employees · Transportation, Logistics, ...  3736179533   
7271   51-200 employees · IT Services and IT Consulting  3731006178   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https:/

In [23]:
df_4.shape

(3542, 9)

In [24]:
df_4_uncomplete_desc_IDs=[]
for i in range(0,len(df_4.iloc[:])):
    if len(df_4.iloc[i]['Job Desception'])<20:
        df_4_uncomplete_desc_IDs.append(df_4.iloc[i]['Job ID'])    

In [25]:
df_4_uncomplete_desc_IDs
len(df_4_uncomplete_desc_IDs)

66

In [26]:
df_4_uncomp_desc=df_4[df_4['Job ID'].isin(df_4_uncomplete_desc_IDs)]
df_4_uncomp_desc

Job Title  \
93             Digital Marketing Strategist and Planner   
265                       Software Engineer-Integration   
651                                     DevOps Engineer   
692                             Real Estate Sales Agent   
699                               Google Ads Specialist   
...                                                 ...   
6375  Junior PR Specialist (Remote Internship – Publ...   
6729                             React Native Developer   
6733                                    Finance Manager   
6960                                 IT Project Manager   
7169                      Talent Acquisition Specialist   

                                               Job Info  \
93    Prime Human Resource Services (Pvt) Ltd · Karā...   
265   People (Professional Employers Pvt Ltd) · Isla...   
651   TecBrix · Lahore, Punjab, Pakistan (Remote)  \...   
692   Fajar Real Estate · Lahore, Punjab, Pakistan (...   
699   RT WEB Solutions · Lahore District, Punjab, Pa...   
...                                                 ...   
6375  TalentKompass Deutschland · Islamabad, Islāmāb...   
6729       FabLearner · Pakistan (Remote)  \n4 days ago   
6733  Four Elements Landscaping · Islamabad, Islāmāb...   
6960  S&P Global · Islamabad, Islāmābād, Pakistan (O...   
7169  DPL · Islamabad, Islāmābād, Pakistan  \n4 days...   

                                    Job Type  \
93                                 Full-time   
265                                Full-time   
651                                Full-time   
692                                Full-time   
699                                Full-time   
...                                      ...   
6375                               Full-time   
6729                               Full-time   
6733                               Full-time   
6960                               Full-time   
7169  Hybrid\n\n \nFull-time\n\n \nAssociate   

                                             Job Link_x  \
93        https://www.linkedin.com/jobs/view/3727011545   
265       https://www.linkedin.com/jobs/view/3727002795   
651   https://www.linkedin.com/jobs/view/3727012776/...   
692   https://www.linkedin.com/jobs/view/3727005694/...   
699   https://www.linkedin.com/jobs/view/3727058923/...   
...                                                 ...   
6375      https://www.linkedin.com/jobs/view/3729322244   
6729      https://www.linkedin.com/jobs/view/3734243297   
6733      https://www.linkedin.com/jobs/view/3734234918   
6960      https://www.linkedin.com/jobs/view/3736407961   
7169      https://www.linkedin.com/jobs/view/3736472247   

                                              employees      Job ID  \
93                                    10,001+ employees  3727011545   
265                              5,001-10,000 employees  3727002795   
651                                     11-50 employees  3727012776   
692                                     11-50 employees  3727005694   
699                                     11-50 employees  3727058923   
...                                                 ...         ...   
6375          1-10 employees · Human Resources Services  3729322244   
6729                                    11-50 employees  3734243297   
6733  See how you compare to 305 applicants. Reactiv...  3734234918   
6960             10,001+ employees · Financial Services  3736407961   
7169   51-200 employees · IT Services and IT Consulting  3736472247   

                                         Job Link_y Job Desception  \
93    https://www.linkedin.com/jobs/view/3727011545  About the job   
265   https://www.linkedin.com/jobs/view/3727002795  About the job   
651   https://www.linkedin.com/jobs/view/3727012776  About the job   
692   https://www.linkedin.com/jobs/view/3727005694  About the job   
699   https://www.linkedin.com/jobs/view/3727058923  About the job   
...                                             ...            ...   

In [27]:
df_5=df_4.drop(df_4_uncomp_desc.index)
df_5

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
7266         Oracle Apps DBA - R12 (Full Customization)   
7267                Engineer, Service-to-Service Fabric   
7268  Technical Product Manager (OTE $80,000/year US...   
7269                               Senior IOS Developer   
7271  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
7266  Rozee.pk · Lahore, Punjab, Pakistan  \n4 days ago   
7267  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
7268  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
7269  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7271  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
7266                        On-site\n\n \nFull-time   
7267  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7268       Remote\n\n \nFull-time\n\n \nEntry level   
7269  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7271       Remote\n\n \nFull-time\n\n \nEntry level   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https://www.linkedin.com/jobs/view/3731001750   
7269  https://www.linkedin.com/jobs/view/3736179533   
7271  https://www.linkedin.com/jobs/view/3731006178   

                                              employees      Job ID  \
0                                       11-50 employees  3727054368   
1                                       11-50 employees  3723395858   
2                                      51-200 employees  3727034665   
3                                       11-50 employees  3727029273   
4                   Mobeen Naqvi is hiring for this job  3727034696   
...                                                 ...         ...   
7266                                501-1,000 employees  3734517227   
7267      5,001-10,000 employees · Software Development  3730554639   
7268   51-200 employees · IT Services and IT Consulting  3731001750   
7269  51-200 employees · Transportation, Logistics, ...  3736179533   
7271   51-200 employees · IT Services and IT Consulting  3731006178   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https:/

In [28]:
3542-66

3476

#### using commen keywords used in IT and Tech

In [29]:
# it_tech_pattern='Software|Computer|Digital|Programming|Data|Freelance|Web|Cloud|Graphic|Game|Information technology|Mobile|Email|Virtual|video|technology|ecommerce|network'
# case=False
# filtered_df = df_5[df_5['Job Desception'].str.contains(it_tech_pattern, case)]
# filtered_df
it_tech_keywords= ['Software', 'Computer', 'Digital', 'Programming', 
                   'Data', 'Freelance', 'Web', 'Cloud', 'Graphic', 
                   'Game', 'Information technology', 'Mobile', 'Email', 
                   'Virtual', 'video', 'technology', 'ecommerce', 'network']
it_tech_keywords
len(it_tech_keywords)

filtered_df_5_desc = df_5[df_5['Job Desception'].apply(lambda desc: any(keyword.lower() in desc.lower() for keyword in it_tech_keywords))]
filtered_df_5_desc

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
7266         Oracle Apps DBA - R12 (Full Customization)   
7267                Engineer, Service-to-Service Fabric   
7268  Technical Product Manager (OTE $80,000/year US...   
7269                               Senior IOS Developer   
7271  Technical Product Manager (OTE $80,000/year US...   

                                               Job Info  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2     Murkez Technologies · Lahore, Punjab, Pakistan...   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                 ...   
7266  Rozee.pk · Lahore, Punjab, Pakistan  \n4 days ago   
7267  Atlassian · APAC Reposted  \n3 days ago\n\n · ...   
7268  Sparkrock · South Asia  \n4 days ago\n\n · 144...   
7269  ekar · Pakistan  \n4 days ago\n\n · 223 applic...   
7271  Sparkrock · South Asia  \n4 days ago\n\n · 137...   

                                           Job Type  \
0                                         Full-time   
1                                         Full-time   
2                                         Full-time   
3                                         Full-time   
4                                         Full-time   
...                                             ...   
7266                        On-site\n\n \nFull-time   
7267  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7268       Remote\n\n \nFull-time\n\n \nEntry level   
7269  Remote\n\n \nFull-time\n\n \nMid-Senior level   
7271       Remote\n\n \nFull-time\n\n \nEntry level   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https://www.linkedin.com/jobs/view/3730554639   
7268  https://www.linkedin.com/jobs/view/3731001750   
7269  https://www.linkedin.com/jobs/view/3736179533   
7271  https://www.linkedin.com/jobs/view/3731006178   

                                              employees      Job ID  \
0                                       11-50 employees  3727054368   
1                                       11-50 employees  3723395858   
2                                      51-200 employees  3727034665   
3                                       11-50 employees  3727029273   
4                   Mobeen Naqvi is hiring for this job  3727034696   
...                                                 ...         ...   
7266                                501-1,000 employees  3734517227   
7267      5,001-10,000 employees · Software Development  3730554639   
7268   51-200 employees · IT Services and IT Consulting  3731001750   
7269  51-200 employees · Transportation, Logistics, ...  3736179533   
7271   51-200 employees · IT Services and IT Consulting  3731006178   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
7266  https://www.linkedin.com/jobs/view/3734517227   
7267  https:/

In [30]:
filtered_df_5_desc['Job Title'].unique()

array(['LinkedIn Data Extractor', 'It Management Intern',
       'SQL Database Developer', ...,
       'Engineer, Service-to-Service Fabric',
       'Technical Product Manager (OTE $80,000/year USD), Sparkrock',
       'Senior IOS Developer'], dtype=object)

In [31]:
len(filtered_df_5_desc['Job Title'].unique()),len(df_3['Job Title'].unique())

(1912, 2343)

### dealing Job titles where jobs desc is abnormal

In [32]:
len(df_3)-len(df_5) # there are 97 rows where Jobs desc= abnormal

97

##### approach:1

In [33]:
df_3.drop(df_5.index)

Job Title  \
93    Digital Marketing Strategist and Planner   
121                           Research Analyst   
265              Software Engineer-Integration   
651                            DevOps Engineer   
692                    Real Estate Sales Agent   
...                                        ...   
6729                    React Native Developer   
6733                           Finance Manager   
6960                        IT Project Manager   
6968                          Graphic Designer   
7169             Talent Acquisition Specialist   

                                               Job Info  \
93    Prime Human Resource Services (Pvt) Ltd · Karā...   
121   Indian School of Business · Hyderabad, Sindh, ...   
265   People (Professional Employers Pvt Ltd) · Isla...   
651   TecBrix · Lahore, Punjab, Pakistan (Remote)  \...   
692   Fajar Real Estate · Lahore, Punjab, Pakistan (...   
...                                                 ...   
6729       FabLearner · Pakistan (Remote)  \n4 days ago   
6733  Four Elements Landscaping · Islamabad, Islāmāb...   
6960  S&P Global · Islamabad, Islāmābād, Pakistan (O...   
6968  Voyagersol · Islamabad, Islāmābād, Pakistan (O...   
7169  DPL · Islamabad, Islāmābād, Pakistan  \n4 days...   

                                    Job Type  \
93                                 Full-time   
121             Full-time · Mid-Senior level   
265                                Full-time   
651                                Full-time   
692                                Full-time   
...                                      ...   
6729                               Full-time   
6733                               Full-time   
6960                               Full-time   
6968                               Full-time   
7169  Hybrid\n\n \nFull-time\n\n \nAssociate   

                                             Job Link_x  \
93        https://www.linkedin.com/jobs/view/3727011545   
121       https://www.linkedin.com/jobs/view/3723336007   
265       https://www.linkedin.com/jobs/view/3727002795   
651   https://www.linkedin.com/jobs/view/3727012776/...   
692   https://www.linkedin.com/jobs/view/3727005694/...   
...                                                 ...   
6729      https://www.linkedin.com/jobs/view/3734243297   
6733      https://www.linkedin.com/jobs/view/3734234918   
6960      https://www.linkedin.com/jobs/view/3736407961   
6968      https://www.linkedin.com/jobs/view/3734595238   
7169      https://www.linkedin.com/jobs/view/3736472247   

                                              employees      Job ID  \
93                                    10,001+ employees  3727011545   
121              501-1,000 employees · Higher Education  3723336007   
265                              5,001-10,000 employees  3727002795   
651                                     11-50 employees  3727012776   
692                                     11-50 employees  3727005694   
...                                                 ...         ...   
6729                                    11-50 employees  3734243297   
6733  See how you compare to 305 applicants. Reactiv...  3734234918   
6960             10,001+ employees · Financial Services  3736407961   
6968                                   51-200 employees  3734595238   
7169   51-200 employees · IT Services and IT Consulting  3736472247   

                                         Job Link_y Job Desception  \
93    https://www.linkedin.com/jobs/view/3727011545  About the job   
121   https://www.linkedin.com/jobs/view/3723336007            NaN   
265   https://www.linkedin.com/jobs/view/3727002795  About the job   
651   https://www.linkedin.com/jobs/view/3727012776  About the job   
692   https://www.linkedin.com/jobs/view/3727005694  About the job   
...                                             ...            ...   
6729  https://www.linkedin.com/jobs/view/3734243297  About the job   
6733  https://www.linkedin.c

##### Approach:2

In [34]:
# FUTURE TASK
# df_3[df_3['Job ID'].isin(df_4_uncomplete_desc_IDs+null_desc_iDS)]

#### getting df_3_abnormal

In [35]:
df_3_abnormal=df_3.drop(df_5.index)
df_3_abnormal

Job Title  \
93    Digital Marketing Strategist and Planner   
121                           Research Analyst   
265              Software Engineer-Integration   
651                            DevOps Engineer   
692                    Real Estate Sales Agent   
...                                        ...   
6729                    React Native Developer   
6733                           Finance Manager   
6960                        IT Project Manager   
6968                          Graphic Designer   
7169             Talent Acquisition Specialist   

                                               Job Info  \
93    Prime Human Resource Services (Pvt) Ltd · Karā...   
121   Indian School of Business · Hyderabad, Sindh, ...   
265   People (Professional Employers Pvt Ltd) · Isla...   
651   TecBrix · Lahore, Punjab, Pakistan (Remote)  \...   
692   Fajar Real Estate · Lahore, Punjab, Pakistan (...   
...                                                 ...   
6729       FabLearner · Pakistan (Remote)  \n4 days ago   
6733  Four Elements Landscaping · Islamabad, Islāmāb...   
6960  S&P Global · Islamabad, Islāmābād, Pakistan (O...   
6968  Voyagersol · Islamabad, Islāmābād, Pakistan (O...   
7169  DPL · Islamabad, Islāmābād, Pakistan  \n4 days...   

                                    Job Type  \
93                                 Full-time   
121             Full-time · Mid-Senior level   
265                                Full-time   
651                                Full-time   
692                                Full-time   
...                                      ...   
6729                               Full-time   
6733                               Full-time   
6960                               Full-time   
6968                               Full-time   
7169  Hybrid\n\n \nFull-time\n\n \nAssociate   

                                             Job Link_x  \
93        https://www.linkedin.com/jobs/view/3727011545   
121       https://www.linkedin.com/jobs/view/3723336007   
265       https://www.linkedin.com/jobs/view/3727002795   
651   https://www.linkedin.com/jobs/view/3727012776/...   
692   https://www.linkedin.com/jobs/view/3727005694/...   
...                                                 ...   
6729      https://www.linkedin.com/jobs/view/3734243297   
6733      https://www.linkedin.com/jobs/view/3734234918   
6960      https://www.linkedin.com/jobs/view/3736407961   
6968      https://www.linkedin.com/jobs/view/3734595238   
7169      https://www.linkedin.com/jobs/view/3736472247   

                                              employees      Job ID  \
93                                    10,001+ employees  3727011545   
121              501-1,000 employees · Higher Education  3723336007   
265                              5,001-10,000 employees  3727002795   
651                                     11-50 employees  3727012776   
692                                     11-50 employees  3727005694   
...                                                 ...         ...   
6729                                    11-50 employees  3734243297   
6733  See how you compare to 305 applicants. Reactiv...  3734234918   
6960             10,001+ employees · Financial Services  3736407961   
6968                                   51-200 employees  3734595238   
7169   51-200 employees · IT Services and IT Consulting  3736472247   

                                         Job Link_y Job Desception  \
93    https://www.linkedin.com/jobs/view/3727011545  About the job   
121   https://www.linkedin.com/jobs/view/3723336007            NaN   
265   https://www.linkedin.com/jobs/view/3727002795  About the job   
651   https://www.linkedin.com/jobs/view/3727012776  About the job   
692   https://www.linkedin.com/jobs/view/3727005694  About the job   
...                                             ...            ...   
6729  https://www.linkedin.com/jobs/view/3734243297  About the job   
6733  https://www.linkedin.c

In [36]:
df_3_abnormal['Job Title']

93      Digital Marketing Strategist and Planner
121                             Research Analyst
265                Software Engineer-Integration
651                              DevOps Engineer
692                      Real Estate Sales Agent
                          ...                   
6729                      React Native Developer
6733                             Finance Manager
6960                          IT Project Manager
6968                            Graphic Designer
7169               Talent Acquisition Specialist
Name: Job Title, Length: 97, dtype: object

In [37]:
df_3_abnormal['Job Title'].unique()

array(['Digital Marketing Strategist and Planner', 'Research Analyst',
       'Software Engineer-Integration', 'DevOps Engineer',
       'Real Estate Sales Agent', 'Software Quality Assurance Analyst',
       'Google Ads Specialist', 'Android Developer',
       'EY Ford Rhodes: Indirect & Direct Tax Compliance & Reporting Lead',
       'Senior Asset Integrity Engineer',
       'Senior .net core API developer', 'Senior MERN Stack Developer',
       '2D Animators', 'Mechanical/Lifting Equipment Inspector',
       'Unity  Android Mobile Game Developer', 'Junior iOS Developer',
       'Executive Chef', 'Senior Angular Developer', 'QHSE Analyst',
       'Digital Marketing Executive (Internship leading to permanent job)',
       'Tax Intern', 'Business Development training program',
       'Finance & Accounts Executive - R2R',
       'Business Development Consultant',
       'Associate, Customer Success (Sales)', 'Treasury Analyst',
       'Admissions Officer', 'Internal Auditor',
       'Ca

In [38]:
conf_titles=['Senior Asset Integrity Engineer','Business Development training program','HARO Link Builder',
             'Project Manager','CGI /Character Artist','Creative Marketing Specialist','Embedded System Engineer',
             'Enterprise Resources Planning Developer','HR Site Lead (Jamshoro)','Senior Care Quality Assurance Associate',
             'Bid Manager','Chief Technology Officer','B2B Account Executive (Sales)',
             'Technical Associate Program 2023 - Excellent Opportunity for Fresh Graduates in Pakistan']

In [39]:
# i am looking at Industry and company
df_3_abnormal[df_3_abnormal['Job Title'].isin(conf_titles)]

Job Title  \
814                     Senior Asset Integrity Engineer   
884               Business Development training program   
1106                                  HARO Link Builder   
1285                                    Project Manager   
1288                              CGI /Character Artist   
1914                      Creative Marketing Specialist   
2044                           Embedded System Engineer   
2061            Enterprise Resources Planning Developer   
2271                            HR Site Lead (Jamshoro)   
3182            Senior Care Quality Assurance Associate   
3736                                        Bid Manager   
5142                           Chief Technology Officer   
5737                      B2B Account Executive (Sales)   
6075  Technical Associate Program 2023 - Excellent O...   

                                               Job Info  \
814   Abyss Solutions Ltd · Islamabad, Islāmābād, Pa...   
884   Tech BPO (Private) Limited · Lahore, Punjab, P...   
1106  HiredSupport · Lahore, Punjab, Pakistan (On-si...   
1285  Raccoon Tech · Islamabad, Islāmābād, Pakistan ...   
1288  Marketing Services · Islamabad, Islāmābād, Pak...   
1914  Rici Melion · Lahore, Punjab, Pakistan (On-sit...   
2044  Jaffer Business Systems · Karāchi, Sindh, Paki...   
2061  Beanbend · Rawalpindi, Punjab, Pakistan (Hybri...   
2271  Haleon · Jamshoro, Sindh, Pakistan (On-site) R...   
3182  Careem · Karāchi, Sindh, Pakistan Reposted  \n...   
3736  Humantek · Karāchi, Sindh, Pakistan (On-site) ...   
5142  Giftkarte · Karāchi, Sindh, Pakistan (On-site)...   
5737  ontoedge · Rawalpindi, Punjab, Pakistan (On-si...   
6075  Siemens · Karāchi, Sindh, Pakistan (Hybrid)  \...   

                          Job Type  \
814   Full-time · Mid-Senior level   
884                      Full-time   
1106                     Full-time   
1285                     Full-time   
1288                     Full-time   
1914                     Full-time   
2044                     Full-time   
2061                     Full-time   
2271  Full-time · Mid-Senior level   
3182  Full-time · Mid-Senior level   
3736                     Full-time   
5142                     Full-time   
5737                     Full-time   
6075       Full-time · Entry level   

                                             Job Link_x  \
814   https://www.linkedin.com/jobs/view/3727033391/...   
884       https://www.linkedin.com/jobs/view/3727542089   
1106      https://www.linkedin.com/jobs/view/3727396682   
1285      https://www.linkedin.com/jobs/view/3727544104   
1288      https://www.linkedin.com/jobs/view/3727520388   
1914      https://www.linkedin.com/jobs/view/3728221292   
2044      https://www.linkedin.com/jobs/view/3728221394   
2061      https://www.linkedin.com/jobs/view/3728213358   
2271      https://www.linkedin.com/jobs/view/3708206982   
3182      https://www.linkedin.com/jobs/view/3709948393   
3736      https://www.linkedin.com/jobs/view/3729936548   
5142      https://www.linkedin.com/jobs/view/3732206096   
5737      https://www.linkedin.com/jobs/view/3732810459   
6075      https://www.linkedin.com/jobs/view/3729082995   

                                              employees      Job ID  \
814   51-200 employees · Automation Machinery Manufa...  3727033391   
884                                    51-200 employees  3727542089   
1106                                   51-200 employees  3727396682   
1285                                    11-50 employees  3727544104   
1288                                    11-50 employees  3727520388   
1914                                501-1,000 employees  3728221292   
2044                                  201-500 employees  3728221394   
2061              Irtaza B. Raja is hiring for this job  3728213358   
2271              10,001+ employees · Consumer Services  3708206982   
3182  1,001-5,000 employees · IT Services and IT Con...  3709948393   
3736                                

#### developing IT tech pattern for Filtering job titles in df_3_abnormal:

In [40]:
# it_tech_pattern1='Digital|software|Analyst|DevOps|Google|Android|2D|Developer|Development|Graphic|Game'
# it_tech_pattern2='|Search Engine Optimization|Creative|Video|Virtual|IT Sales|Application|Technical Content Writer'
# it_tech_pattern3='|IT Project|Quality Assurance|Project'

# it_tech_pattern=it_tech_pattern1+it_tech_pattern2+it_tech_pattern3
# it_tech_pattern

# case=False
# filtered_df = df_3_abnormal[df_3_abnormal['Job Title'].str.contains(it_tech_pattern, case)]
# filtered_df
# len(filtered_df)

it_tech_pattern1 = ['Digital', 'software', 'Analyst', 'DevOps', 'Google', 'Android', '2D', 'Developer', 'Development', 'Graphic', 'Game']
it_tech_pattern2 = ['Search Engine Optimization', 'Creative', 'Video', 'Virtual', 'IT Sales', 'Application', 'Technical Content Writer']
it_tech_pattern3 = ['IT Project', 'Quality Assurance', 'Project']

it_tech_keywords = it_tech_pattern1 + it_tech_pattern2 + it_tech_pattern3
it_tech_keywords
len(it_tech_keywords)

# it_tech_keywords = ['Digital', 'software', 'Analyst', 'DevOps', 'Google', 'Android', '2D', 'Developer', 'Development', 'Graphic', 'Game',
#                     'Search Engine Optimization', 'Creative', 'Video', 'Virtual', 'IT Sales', 'Application', 'Technical Content Writer',
#                     'IT Project', 'Quality Assurance', 'Project']
# it_tech_keywords
# len(it_tech_keywords)

filtered_df = df_3_abnormal[df_3_abnormal['Job Title'].apply(lambda title: any(keyword.lower() in title.lower() for keyword in it_tech_keywords))]
filtered_df
len(filtered_df)

57

#### Filtering Tech job titles in df_3_abnormal:

In [41]:
it_tech_keywords = ['Digital', 'software', 'Analyst', 'DevOps', 'Google', 'Android', '2D', 'Developer', 'Development', 'Graphic', 'Game',
                    'Search Engine Optimization', 'Creative', 'Video', 'Virtual', 'IT Sales', 'Application', 'Technical Content Writer',
                    'IT Project', 'Quality Assurance', 'Project']
it_tech_keywords
len(it_tech_keywords)

filtered_df_3_abnormal = df_3_abnormal[df_3_abnormal['Job Title'].apply(lambda title: any(keyword.lower() in title.lower() for keyword in it_tech_keywords))]
filtered_df_3_abnormal
# len(ffiltered_df_3_abnormal)

Job Title  \
93             Digital Marketing Strategist and Planner   
121                                    Research Analyst   
265                       Software Engineer-Integration   
651                                     DevOps Engineer   
694                  Software Quality Assurance Analyst   
699                               Google Ads Specialist   
703                                   Android Developer   
815                      Senior .net core API developer   
816                         Senior MERN Stack Developer   
819                                        2D Animators   
821                Unity  Android Mobile Game Developer   
822                                Junior iOS Developer   
864                            Senior Angular Developer   
880                                        QHSE Analyst   
881   Digital Marketing Executive (Internship leadin...   
884               Business Development training program   
893                     Business Development Consultant   
914                                    Treasury Analyst   
922                                Business Development   
928   Upwork and Fiverr Business Development Specialist   
931                                    Graphic Designer   
936   AI-Augmented Software Developer, Skyvera (Remo...   
942                     Business Development Specialist   
1021                                       Game Planner   
1089          Principal Software Engineer (khi/lhr/Isl)   
1090                         Sales Business Development   
1091                          Digital Marketing Manager   
1092               Search Engine Optimization Team Lead   
1285                                    Project Manager   
1286                                Front-End Developer   
1291                                  Creative Designer   
1293                                      PHP Developer   
1296  Back End Developer, Trilogy (Remote) - $100,00...   
1908                                 Ai Video Generator   
1912                       Bussiness Development Intern   
1913                     Business Development Executive   
1914                      Creative Marketing Specialist   
2061            Enterprise Resources Planning Developer   
2364                      Senior React Native Developer   
2426                          Web Developer - WordPress   
2640             Virtual Customer Experience Internship   
2993  Principal Software Engineer, GFI Software (Rem...   
3182            Senior Care Quality Assurance Associate   
3534                               Senior Web Developer   
4264                       Business Development Manager   
4274                                 IT Sales Executive   
4461   Backend Developer (Custom WordPress and Laravel)   
4486  Senior Full-stack Java / Angular Developer - +...   
4660                                 Google Ads Manager   
4715                           Creative Unity Developer   
5140                       Application Support Engineer   
5308                    System SupportAnalyst (Finacle)   
6087  Junior Market Analyst (Remote Internship – Mar...   
6598                           Technical Content Writer   
6729                             React Native Developer   
6960                                 IT Project Manager   
6968                                   Graphic Designer   

                                               Job Info  \
93    Prime Human Resource Services (Pvt) Ltd · Karā...   
121   Indian School of Business · Hyderabad, Sindh, ...   
265   People (Professional Employers Pvt Ltd) · Isla...   
651   TecBrix · Lahore, Punjab, Pakistan (Remote)  \...   
694   Stewart Pakistan Private Limited · Lahore, Pun...   
699   RT WEB Solutions · Lahore District, Punjab, Pa...   
703   MWAN Mobile · Islāmābād, Pakistan (On-site)  \...   
815   MaxRemind Inc · Islamabad, Islāmābād, Pakistan...   
816   HR Ways - Hiring Tech Talent · Lahore, Punjab,...   
819   Rex Technologies · Lahore D

### concatenating filtered_df_3_abnormal and filtered_df_5_desc:

In [42]:
filtered_df_3_Tech = pd.concat([filtered_df_5_desc,filtered_df_3_abnormal], ignore_index=True)
filtered_df_3_Tech

Job Title  \
0                               LinkedIn Data Extractor   
1                                  It Management Intern   
2                                SQL Database Developer   
3                       Business Intelligence Developer   
4                            Web Developer - Internship   
...                                                 ...   
2892  Junior Market Analyst (Remote Internship – Mar...   
2893                           Technical Content Writer   
2894                             React Native Developer   
2895                                 IT Project Manager   
2896                                   Graphic Designer   

                                               Job Info   Job Type  \
0           EcomFleet · Pakistan (Remote)  \n1 week ago  Full-time   
1     Digicon Valley · Lahore, Punjab, Pakistan (On-...  Full-time   
2     Murkez Technologies · Lahore, Punjab, Pakistan...  Full-time   
3     DevNatives · Islamabad, Islāmābād, Pakistan (H...  Full-time   
4     ItCreaters · Lahore, Punjab, Pakistan (On-site...  Full-time   
...                                                 ...        ...   
2892  TalentKompass Deutschland · Lahore, Punjab, Pa...  Full-time   
2893  Logico · Faisalabad District, Punjab, Pakistan...  Full-time   
2894       FabLearner · Pakistan (Remote)  \n4 days ago  Full-time   
2895  S&P Global · Islamabad, Islāmābād, Pakistan (O...  Full-time   
2896  Voyagersol · Islamabad, Islāmābād, Pakistan (O...  Full-time   

                                         Job Link_x  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
2892  https://www.linkedin.com/jobs/view/3729320026   
2893  https://www.linkedin.com/jobs/view/3733479114   
2894  https://www.linkedin.com/jobs/view/3734243297   
2895  https://www.linkedin.com/jobs/view/3736407961   
2896  https://www.linkedin.com/jobs/view/3734595238   

                                      employees      Job ID  \
0                               11-50 employees  3727054368   
1                               11-50 employees  3723395858   
2                              51-200 employees  3727034665   
3                               11-50 employees  3727029273   
4           Mobeen Naqvi is hiring for this job  3727034696   
...                                         ...         ...   
2892  1-10 employees · Human Resources Services  3729320026   
2893                            11-50 employees  3733479114   
2894                            11-50 employees  3734243297   
2895     10,001+ employees · Financial Services  3736407961   
2896                           51-200 employees  3734595238   

                                         Job Link_y  \
0     https://www.linkedin.com/jobs/view/3727054368   
1     https://www.linkedin.com/jobs/view/3723395858   
2     https://www.linkedin.com/jobs/view/3727034665   
3     https://www.linkedin.com/jobs/view/3727029273   
4     https://www.linkedin.com/jobs/view/3727034696   
...                                             ...   
2892  https://www.linkedin.com/jobs/view/3729320026   
2893  https://www.linkedin.com/jobs/view/3733479114   
2894  https://www.linkedin.com/jobs/view/3734243297   
2895  https://www.linkedin.com/jobs/view/3736407961   
2896  https://www.linkedin.com/jobs/view/3734595238   

                                         Job Desception  \
0     About the job\n            \n\n \nAbout Us:We ...   
1     About the job\n            \n\n \nWe are looki...   
2     About the job\n            \n\n \nCompany Over...   
3     About the job\n            \n\n \nBI Developer...   
4     About the job\n            \n\n \nIt is a prog...   
...                                                 ...   
2892

In [43]:
2840+57

2897

### getting non_tech_df3

In [44]:
# df_3.drop(filtered_df_3_Tech.index)
# # KeyError 

In [45]:
non_tech_df3=df_3[~df_3['Job ID'].isin(filtered_df_3_Tech['Job ID'])]
non_tech_df3

Job Title  \
5     Research Analyst for Business Subjects (Office...   
6                          Performance Marketing Intern   
94                 Search Engine Optimization Executive   
96                            Machine Learning Engineer   
106                             Human Resources Trainee   
...                                                 ...   
7244    Deputy Chief Engineer (Mechanical - Gas Engine)   
7250                  Deputy Chief Manager (Compliance)   
7254                                         Prosecutor   
7258                                 Manager - Taxation   
7262                  Chief Manager (Board Secretariat)   

                                               Job Info  \
5     TechoMatrix · Karāchi, Sindh, Pakistan (On-sit...   
6     8th Loop · Karāchi, Sindh, Pakistan (On-site) ...   
94    Brandix Soft  · Lahore, Punjab, Pakistan (On-s...   
96    Mexa Solutions · Lahore, Punjab, Pakistan (On-...   
106   TechSwivel (Private) Limited · Lahore, Punjab,...   
...                                                 ...   
7244  Sui Southern Gas Company Limited · Karachi Div...   
7250  Sui Southern Gas Company Limited · Karachi Div...   
7254  Sui Southern Gas Company Limited · Quetta, Bal...   
7258  Sui Southern Gas Company Limited · Karachi Div...   
7262  Sui Southern Gas Company Limited · Karachi Div...   

                                            Job Type  \
5                                          Full-time   
6                                          Full-time   
94                                         Full-time   
96                                         Full-time   
106                                        Full-time   
...                                              ...   
7244          On-site\n\n \nFull-time\n\n \nDirector   
7250           Hybrid\n\n \nFull-time\n\n \nDirector   
7254  On-site\n\n \nFull-time\n\n \nMid-Senior level   
7258   Hybrid\n\n \nFull-time\n\n \nMid-Senior level   
7262           Hybrid\n\n \nFull-time\n\n \nDirector   

                                         Job Link_x  \
5     https://www.linkedin.com/jobs/view/3727008042   
6     https://www.linkedin.com/jobs/view/3727056183   
94    https://www.linkedin.com/jobs/view/3727025862   
96    https://www.linkedin.com/jobs/view/3727037227   
106   https://www.linkedin.com/jobs/view/3726899313   
...                                             ...   
7244  https://www.linkedin.com/jobs/view/3735935876   
7250  https://www.linkedin.com/jobs/view/3735942016   
7254  https://www.linkedin.com/jobs/view/3735937615   
7258  https://www.linkedin.com/jobs/view/3735942023   
7262  https://www.linkedin.com/jobs/view/3735942017   

                               employees      Job ID  \
5                        11-50 employees  3727008042   
6                        11-50 employees  3727056183   
94                      51-200 employees  3727025862   
96                       11-50 employees  3727037227   
106                      11-50 employees  3726899313   
...                                  ...         ...   
7244  5,001-10,000 employees · Utilities  3735935876   
7250  5,001-10,000 employees · Utilities  3735942016   
7254  5,001-10,000 employees · Utilities  3735937615   
7258  5,001-10,000 employees · Utilities  3735942023   
7262  5,001-10,000 employees · Utilities  3735942017   

                                         Job Link_y  \
5     https://www.linkedin.com/jobs/view/3727008042   
6     https://www.linkedin.com/jobs/view/3727056183   
94    https://www.linkedin.com/jobs/view/3727025862   
96    https://www.linkedin.com/jobs/view/3727037227   
106   https://www.linkedin.com/jobs/view/3726899313   
...                                             ...   
7244  https://www.linkedin.com/jobs/view/3735935876   
7250  https://www.linkedin.com/jobs/view/3735942016   
7254  https://www.linkedin.com/jobs/view/3735937615   
7258  https://www.linkedin.com/jobs/view/3735942023   
7262  https://ww

In [46]:
2897+676

3573

### analyzing filtered_df_3_Tech['Job Title'] and non_tech_df3['Job Title']

In [47]:
len(filtered_df_3_Tech['Job Title'].unique()),len(df_3['Job Title'].unique()),len(non_tech_df3['Job Title'].unique())

(1946, 2343, 505)

***why there are 505 unique job titles in non_tech_df3 it should be 397 so It may be unique titles in IT and non IT jobs(mixing)***

In [48]:
1946+505

2451

In [49]:
len(df_3['Job Title'].unique())-len(filtered_df_3_Tech['Job Title'].unique())

397

In [50]:
505-397
# 2451-2343

108

***108 job titles are mixed***

### finding commen job titles

In [51]:
# l1=['A','B','C','D']
# l2=['A','B','C']
# l2 in l1
# type(l1)

In [52]:
commen_job_titles=np.intersect1d(filtered_df_3_Tech['Job Title'].unique(),non_tech_df3['Job Title'].unique())
# type(commen_job_titles)
commen_job_titles
# len(commen_job_titles)

array(['Account Assistant', 'Account Executive', 'Accountant',
       'Admin Officer', 'Admissions Officer', 'Android Developer',
       'Angular Developer', 'Animator', 'Appointment Setter',
       'Assistant Manager', 'Assistant Manager Sales',
       'Assistant Sales Manager', 'Associate', 'Bid Manager',
       'Bid Specialist', 'Business Development Associate',
       'Business Development Executive', 'Business Development Manager',
       'Business Development Officer',
       'Business Development Representative',
       'Business Development Specialist', 'Call Center Representative',
       'Career Counseling Specialist', 'Chief Technology Officer',
       'Civil Engineer', 'Closer',
       'Computer Numerical Control Programmer', 'Content Writer',
       'Customer Sales Representative', 'Customer Service Officer',
       'Customer Service Representative',
       'Customer Support Representative', 'Digital Marketing Manager',
       'Digital Marketing Specialist', 'Educational C

***There are some Tech jobs in non_tech_df and non_tech_jobs in tech_df***

In [53]:
IT_titles_in_non_tech_df3=['Android Developer','Angular Developer','Animator','Bid Manager','Closer',
                           'Computer Numerical Control Programmer','Digital Marketing Manager',
                           'Digital Marketing Specialist','Flutter Developer','Machine Learning Engineer',
                           'Senior Web Developer','Shopify Developer','Social Media Manager','Sr. Angular Developer',
                           'Technical Content Writer','Technical Project Manager','UI/UX Designer','Upwork Bidder',
                           'Virtual Customer Experience Internship','Wordpress Developer','iOS Developer']
IT_titles_in_non_tech_df3

['Android Developer',
 'Angular Developer',
 'Animator',
 'Bid Manager',
 'Closer',
 'Computer Numerical Control Programmer',
 'Digital Marketing Manager',
 'Digital Marketing Specialist',
 'Flutter Developer',
 'Machine Learning Engineer',
 'Senior Web Developer',
 'Shopify Developer',
 'Social Media Manager',
 'Sr. Angular Developer',
 'Technical Content Writer',
 'Technical Project Manager',
 'UI/UX Designer',
 'Upwork Bidder',
 'Virtual Customer Experience Internship',
 'Wordpress Developer',
 'iOS Developer']

In [54]:
filtered_df_3_Tech['Job Title'].unique()[:100]

array(['LinkedIn Data Extractor', 'It Management Intern',
       'SQL Database Developer', 'Business Intelligence Developer',
       'Web Developer - Internship', 'Sr. SDET Engineer',
       'Associate Engineer', 'Fresh Graduate Engineers',
       'Fraud Monitoring Officer', 'Business System Analyst',
       'Research Analyst - Finance', 'HR Intern',
       'Alfa Origin is looking for sales and HR Intern',
       'Sales Operations Specialist', 'ASO Internee',
       'Trainee R&D Engineer', 'Dotnet+React Developer',
       'Business Development Executive (REMOTE) up to A$2,500/month',
       'Artificial Intelligence Developer', 'Web Designer',
       'Contact Centre Rep (REMOTE) up to A$ 2,000/month',
       'Software Engineer', 'SQL Database Engineer',
       'Junior Business Developer', 'Human Resources Intern',
       'Antifraud Specialist', 'Management Trainee Officer',
       'Jr Business Developer (Upwork)',
       'Software Engineer - Product Operation',
       'International Pro

In [56]:
non_tech_df3['Job Title'].unique()[:100]

array(['Research Analyst for Business Subjects (Office Based + Night Shift)',
       'Performance Marketing Intern',
       'Search Engine Optimization Executive',
       'Machine Learning Engineer', 'Human Resources Trainee',
       'Business Development Executive',
       'Fintech B2B Marketing Executive',
       'Customer Support Representative', 'Ecommerce Manager',
       'Territory Incharge [Future roles - Multiple Locations',
       'Upwork Bidding Specialist', 'Upwork Bidder',
       'Student Leader - Chemistry', 'Customer Service Representative',
       'Business Development Specialist', 'Internal Auditor',
       'Business Development Representative-Australia region',
       'Marketing Executive', 'Business Development Representative',
       'Human Resources and Admin Executive', 'Tax Officer',
       'Sales Executive', 'Officer Sales Support - Sukkur',
       'Study Abroad Advisor', 'Customer Sales Representative',
       'Sales Officer', 'Export Operations Specialist',
   

In [55]:
filtered_df_3_Tech['Job Title'].value_counts()[:50]

Job Title
Business Development Executive                                              34
Software Engineer                                                           28
Educational Counselor                                                       26
Sales Executive                                                             23
Senior Software Engineer                                                    20
Business Development Manager                                                16
Graphic Designer                                                            16
Dotnet Developer                                                            14
DevOps Engineer                                                             13
Full Stack Engineer                                                         13
Wordpress Developer                                                         12
Business Development Officer                                                12
PHP Developer                             

##### investigating Educational Counselor job

In [148]:
# filtered_df_3_Tech[filtered_df_3_Tech['Job Title']=='Educational Counselor'].iloc[0]['Job Desception']
# 'software' in filtered_df_3_Tech[filtered_df_3_Tech['Job Title']=='Educational Counselor'].iloc[0]['Job Desception']
# filtered_df_3_Tech[filtered_df_3_Tech['Job Title']=='Educational Counselor'].iloc[0]['Job Info']
# # These are some technical education related jobs

In [57]:
non_tech_df3['Job Title'].value_counts()[:50]

Job Title
Sales Executive                                                     23
Business Development Executive                                      13
Sales Specialist                                                    11
Salesperson                                                          9
Business Development Manager                                         7
Sales And Marketing Specialist                                       7
Customer Service Representative                                      7
Carrier Onboarding / Sales Executive at a Truck Dispatching LLC!     6
Business Development Specialist                                      5
Proofreader                                                          5
Associate Project Sales                                              4
Assistant Manager Sales                                              4
Assistant Manager                                                    4
International Sales Executive                                      